velox/substrait/VeloxToSubstraitType.cpp

```diff
@@ -0,0 +1,336 @@
+      return *sType;
+    }
+    case velox::TypeKind::DOUBLE: {
+      sType->set_allocated_fp64(new ::substrait::Type_FP64());
+      return *sType;
+    }
+    case velox::TypeKind::VARCHAR: {
+      sType->set_allocated_varchar(new ::substrait::Type_VarChar());
+      return *sType;
+    }
+    case velox::TypeKind::VARBINARY: {
+      sType->set_allocated_binary(new ::substrait::Type_Binary());
+      return *sType;
+    }
+    case velox::TypeKind::TIMESTAMP: {
+      sType->set_allocated_timestamp(new ::substrait::Type_Timestamp());
+      return *sType;
+    }
+    case velox::TypeKind::ARRAY: {
+      ::substrait::Type_List* sTList = new ::substrait::Type_List();
+      const std::shared_ptr<const Type> vArrayType =
+          vType->asArray().elementType();
+      ::substrait::Type sListType =
+          toSubstraitType(vArrayType, sTList->mutable_type());
+
+      sType->set_allocated_list(sTList);
+      return *sType;
+    }
+    case velox::TypeKind::MAP: {
+      ::substrait::Type_Map* sMap = new ::substrait::Type_Map();
+      const std::shared_ptr<const Type> vMapKeyType = vType->asMap().keyType();
+      const std::shared_ptr<const Type> vMapValueType =
+          vType->asMap().valueType();
+
+      toSubstraitType(vMapKeyType, sMap->mutable_key());
+      toSubstraitType(vMapValueType, sMap->mutable_value());
+
+      sType->set_allocated_map(sMap);
+      return *sType;
+    }
+    case velox::TypeKind::UNKNOWN:
+    case velox::TypeKind::FUNCTION:
+    case velox::TypeKind::OPAQUE:
+    case velox::TypeKind::INVALID:
+    default:
+      throw std::runtime_error(
+          "Unsupported type " + std::string(vType->kindName()));
+  }
+}
+
+::substrait::NamedStruct* VeloxToSubstraitTypeConvertor::toSubstraitNamedStruct(
+    const velox::RowTypePtr& vRow,
+    ::substrait::NamedStruct* sNamedStruct) {
+  int64_t vSize = vRow->size();
+  std::vector<std::string> vNames = vRow->names();
+  std::vector<std::shared_ptr<const Type>> vTypes = vRow->children();
+
+  for (int64_t i = 0; i < vSize; ++i) {
+    std::string vName = vNames.at(i);
+    std::shared_ptr<const Type> vType = vTypes.at(i);
+    sNamedStruct->add_names(vName);
+    ::substrait::Type* sStruct = sNamedStruct->mutable_struct_()->add_types();
+
+    toSubstraitType(vType, sStruct);
+  }
+
+  return sNamedStruct;
+}
+
+::substrait::Expression_Literal*
+VeloxToSubstraitTypeConvertor::processVeloxValueByType(
+    ::substrait::Expression_Literal_Struct* sLitValue,
+    ::substrait::Expression_Literal* sField,
+    const velox::VectorPtr& children) {
+  // to handle the null value. TODO need to confirm
+  std::optional<vector_size_t> nullCount = children->getNullCount();
+  // should be the same with rowValue->type();
+  std::shared_ptr<const Type> childType = children->type();
+  switch (childType->kind()) {
+    case velox::TypeKind::BOOLEAN: {
+      auto childToFlatVec = children->asFlatVector<bool>();
+      vector_size_t flatVecSzie = childToFlatVec->size();
```

Comment:
typos: flatVecSzie -> size

Does this method expect a vector of size 1 ? If so, let's add VELOX_CHECK_EQ(size, 1) and remove the loop. Otherwise, please, explain what this method is used for. If you need to loop over values and check for nulls, you can do it like so:

```
for (int64_t i = 0; i < size; i++) {
   if (vector->IsNullAt(i)) {
     // process null
   } else {
     bool value = vector->valueAt(i);
   }
```

Reply:
This method is used to get the batch size in the vector. Add comments here and move this function into  VeloxToSubstraitExpr.cpp cause maybe more reasonable in the Expr Convertor.